sae_lens/training/toy_models.py

```diff
@@ -271,17 +211,138 @@ def optimize(
+        else:
+            return reconstructed
+
+    def calculate_loss(
+        self,
+        out: Float[Tensor, "batch features"],
+        batch: Float[Tensor, "batch features"],
+    ) -> Float[Tensor, ""]:
+        """
+        Calculates the loss for a given batch, using this loss described in the Toy Models paper:
+
+            https://transformer-circuits.pub/2022/toy_model/index.html#demonstrating-setup-loss
+
+        Note, `model.importance` is guaranteed to broadcast with the shape of `out` and `batch`.
+        """
+        error = self.importance * ((batch - out) ** 2)
+        loss = einops.reduce(error, "batch features -> ()", "mean").sum()
+        return loss
+
+
+class ReluOutputModelCE(ReluOutputModel):
+    """
+    A variant of Anthropic's ReLU Output Model.
+    This model is trained with a Cross Entropy loss instead of MSE loss.
+    The model task is to identify which feature has the largest magnitude activation in the input.
+    The model has an extra feature dimension which is set to a constant nonzero value,
+    which allows for proper classification when all features are zero.
+    """
+
+    W: Float[Tensor, "n_hidden n_features"]
+    b_final: Tensor
+    # Our linear map is x -> ReLU(W.T @ W @ x + b_final)
+
+    def __init__(
+        self,
+        cfg: ToyConfig,
+        device: torch.device = torch.device("cpu"),
+        extra_feature_value: float = 1e-6,
+    ):
+        super().__init__(cfg)
+        self.extra_feature_value = extra_feature_value
+
+        self.W = nn.Parameter(
+            nn.init.xavier_normal_(t.empty((cfg.n_hidden, cfg.n_features + 1)))
+        )
+        self.b_final = nn.Parameter(t.zeros((cfg.n_features + 1)))
+        self.to(device)
+
+    def generate_batch(self, batch_size: int) -> Float[Tensor, "batch_size features"]:
+        """Adds an extra feature to the batch, which is set to a constant nonzero value."""
+        batch = super().generate_batch(batch_size)
+        extra_feature = self.extra_feature_value * t.ones((batch_size, 1)).to(
+            batch.device
+        )
+        return t.cat((batch, extra_feature), dim=-1)
+
+    def calculate_loss(
+        self,
+        out: Float[Tensor, "batch features"],
+        batch: Float[Tensor, "batch features"],
+    ) -> Float[Tensor, ""]:
+        """
+        Calculates the loss for a given batch.
+        Loss is calculated using Cross Entropy loss, where the true probability distribution
+        is a one-hot encoding of the feature with the largest magnitude activation in the input.
+        Model outputs (raw logits) are weighted by importance before being passed through CE loss.
+
+        Note, `model.importance` is guaranteed to broadcast with the shape of `out` and `batch`.
+        """
+        max_feat_indices = t.argmax(batch, dim=-1)
+        loss = F.cross_entropy(
+            (self.importance * out).squeeze(), max_feat_indices.squeeze()
+        )
+        return loss
```

Comment:
Here as well, this seems like an import part of the code to have coverage on. But only if you have time - I'm happy to add a test too